src/measure_theory/function/conditional_expectation.lean

```diff
@@ -1266,6 +1319,314 @@ lemma condexp_ind_disjoint_union (hs : measurable_set s) (ht : measurable_set t)
+... = condexp_L1_clm hm μ (hf.to_L1 f) + condexp_L1_clm hm μ (hg.to_L1 g) :
+  (condexp_L1_clm hm μ).map_add _ _
+... = condexp_L1 hm μ f + condexp_L1 hm μ g :
+  by rw [condexp_L1_eq hf, condexp_L1_eq hg]
+
+lemma condexp_L1_neg (f : α → F') :
+  condexp_L1 hm μ (-f) = - condexp_L1 hm μ f :=
+begin
+  by_cases hf : integrable f μ,
+  { calc condexp_L1 hm μ (-f) = condexp_L1_clm hm μ (hf.neg.to_L1 (-f)) : condexp_L1_eq hf.neg
+    ... = condexp_L1_clm hm μ (- hf.to_L1 f) : by rw integrable.to_L1_neg _ hf
+    ... = - condexp_L1_clm hm μ (hf.to_L1 f) : (condexp_L1_clm hm μ).map_neg _
+    ... = - condexp_L1 hm μ f : by rw condexp_L1_eq hf, },
+  { rw [condexp_L1_undef hf, condexp_L1_undef (mt integrable_neg_iff.mp hf), neg_zero], },
+end
+
+lemma condexp_L1_smul (c : 𝕜) (f : α → F') :
+  condexp_L1 hm μ (c • f) = c • condexp_L1 hm μ f :=
+begin
+  by_cases hf : integrable f μ,
+  { calc condexp_L1 hm μ (c • f) = condexp_L1_clm hm μ ((hf.smul c).to_L1 (c • f)) :
+      condexp_L1_eq (hf.smul c)
+    ... = condexp_L1_clm hm μ (c • hf.to_L1 f) : by rw integrable.to_L1_smul' _ hf c
+    ... = c • condexp_L1_clm hm μ (hf.to_L1 f) : condexp_L1_clm_smul c (hf.to_L1 f)
+    ... = c • condexp_L1 hm μ f : by rw condexp_L1_eq hf, },
+  { by_cases hc : c = 0,
+    { rw [hc, zero_smul, zero_smul, condexp_L1_zero], },
+    rw [condexp_L1_undef hf, condexp_L1_undef (mt (integrable_smul_iff hc f).mp hf), smul_zero], },
+end
+
+lemma condexp_L1_sub (hf : integrable f μ) (hg : integrable g μ) :
+  condexp_L1 hm μ (f - g) = condexp_L1 hm μ f - condexp_L1 hm μ g :=
+by rw [sub_eq_add_neg, sub_eq_add_neg, condexp_L1_add hf hg.neg, condexp_L1_neg g]
+
+end condexp_L1
+
+section condexp
+
+/-! # Conditional expectation of a function -/
+
+open_locale classical
+
+local attribute [instance] fact_one_le_one_ennreal
+
+variables {𝕜} {m m0 : measurable_space α} {μ : measure α} [borel_space 𝕜] [is_scalar_tower ℝ 𝕜 F']
+  {hm : m ≤ m0} [sigma_finite (μ.trim hm)] {f g : α → F'} {s : set α}
+
+/-- Conditional expectation of a function. Its value is 0 if the function is not integrable. -/
+def condexp (hm : m ≤ m0) (μ : measure α) [sigma_finite (μ.trim hm)] (f : α → F') : α → F' :=
+ae_measurable'_condexp_L1.mk (condexp_L1 hm μ f)
+
+notation  μ `[` f `|` hm `]` := condexp hm μ f
+
+lemma condexp_ae_eq_condexp_L1 (f : α → F') : μ[f|hm] =ᵐ[μ] condexp_L1 hm μ f :=
+(ae_measurable'.ae_eq_mk ae_measurable'_condexp_L1).symm
+
+lemma condexp_ae_eq_condexp_L1_clm (hf : integrable f μ) :
+  μ[f|hm] =ᵐ[μ] condexp_L1_clm hm μ (hf.to_L1 f) :=
+begin
+  refine (condexp_ae_eq_condexp_L1 f).trans (eventually_of_forall (λ x, _)),
+  rw condexp_L1_eq hf,
+end
+
+lemma condexp_undef (hf : ¬ integrable f μ) : μ[f|hm] =ᵐ[μ] 0 :=
+begin
+  refine (condexp_ae_eq_condexp_L1 f).trans (eventually_eq.trans _ (coe_fn_zero _ 1 _)),
+  rw condexp_L1_undef hf,
+end
+
+lemma condexp_zero : μ[(0 : α → F')|hm] =ᵐ[μ] 0 :=
```

Comment:
do you think it would be worth tweaking the definition of `ae_measurable'.mk` (or whatever version you use) to make sure it maps `0` to `0` exactly, to avoid almost sure equalities like that when possible?

Reply:
That's not enough to get zero here. `condexp` is `ae_measurable'_condexp_L1.mk (condexp_L1 hm μ f)`, so it's an `ae_measurable'.mk` of `⇑(0 : α →₁[μ] F')`, which may not be the zero function.

Thinking about making a more informed choice for that `mk`, the first thing that comes to my mind is that we may want it to be the identity on `m`-measurable functions, but this is also of no help here, since `⇑(0 : α →₁[μ] F')` may not be `m`-measurable.